### PR TITLE
[Robot Software] Support Variable Chip

### LINKED ISSUE #3012: [Robot Software] Support variable chip power
### Description of the task
Similar to the variable kick power, we should support variable chip power. 

### Acceptance criteria

- [ ] Support variable chip power
- [ ] Collaborate with mech to tune the pulse_width to chip distance relationship
- [ ] Test this using Robot Diagnostics
- [ ] Update comment in Robot Diagnostics that power is constant for chipping

This is needed as part of #3157

### PR DESCRIPTION
**Pull request by Andrewyx**
## Please fill out the following before requesting review on this PR

We can now chip varying distances with different powers up to a current arbitrary distance of 2m. More accurate at mid ranges.

### Testing Done
Robot Diagnostics + testing with ball & tape measure. Calibrated redis values for chipping.

### Resolved Issues
Resolves #3012. Implements Dannon's branch.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/proto/message_translation/power_frame_msg.hpp [hunk 1]
```diff
@@ -146,7 +149,10 @@ TbotsProto_PowerPulseControl inline createNanoPbPowerPulseControl(
         case TbotsProto::PowerControl::ChickerControl::kChipDistanceMeters:
             nanopb_control.chicker.which_chicker_command =
                 TbotsProto_PowerPulseControl_ChickerControl_chip_pulse_width_tag;
-            nanopb_control.chicker.chicker_command.chip_pulse_width = chip_pulse_width;
+            nanopb_control.chicker.chicker_command.chip_pulse_width =
```
Comment (itsarune) [inline src/proto/message_translation/power_frame_msg.hpp]: can you limit the max chip_pulse_width to a safe number (ask elec)

**Review by itsarune — COMMENTED** [review]
looks good, minor safety nit

File: src/proto/message_translation/power_frame_msg.hpp (hunk 1 above)
Comment (Andrewyx) [reply src/proto/message_translation/power_frame_msg.hpp]: It is limited, see:
https://github.com/UBC-Thunderbots/Software/pull/3265/files#diff-45ff27e52591072cc56ef258c891210d55e70f0cddfa1c5415691c0b7684f1f6R136-R137

This value can be tuned further though.

Comment (Andrewyx) [reply src/proto/message_translation/power_frame_msg.hpp]: Actually I thought of a better way to limit it by moving the min into the switch case directly, Will fix

Comment (Andrewyx) [reply src/proto/message_translation/power_frame_msg.hpp]: fixed

File: src/software/ai/hl/stp/primitive/primitive_test.cpp [hunk 2]
```diff
@@ -19,7 +19,7 @@ class PrimitiveTest : public testing::Test
     RobotConstants_t robot_constants = create2021RobotConstants();
     Robot robot                      = TestUtil::createRobotAtPos(Point(0, 0));
     std::shared_ptr<World> world     = TestUtil::createBlankTestingWorld();
-    RobotNavigationObstacleFactory obstacle_factory =
+    RobotNavigationObstacleFactory obstacle_factory =a
```
Comment (itsarune) [inline src/software/ai/hl/stp/primitive/primitive_test.cpp:22]: uh oh

**Comment by itsarune** [comment]
Also seems like the code doesn't compile

File: src/shared/constants.h [hunk 3]
```diff
@@ -184,6 +184,11 @@ static const unsigned DIV_B_NUM_ROBOTS = 6;
 static const int MAX_KICK_CONSTANT       = 850;
 static const double MAX_KICK_COEFFICIENT = 0.4;
 
+// Chip Distance to Pulse Width Safety Constraint Constants
+// TODO: UPDATE CONSTANTS ONCE MORE CHIP DATA IS TAKEN FOR ROBOTS
+static const uint32_t MAX_CHIP_PULSE_WIDTH = 7500;
```
Comment (itsarune) [inline src/shared/constants.h:189]: you need to `#include <cstdint>` or just use an int and `static_cast` later since you're already doing it

**Review by PhilD71 — COMMENTED** [review]
Looks good to me pending arun's comments. Saw this in person and its super cool, nice work!

**Comment by itsarune** [comment]
@Andrewyx Could you clean this up, retarget it for master and re-open this PR? We're going to keep #3012 open.

**Comment by Andrewyx** [comment]
> @Andrewyx Could you clean this up, retarget it for master and re-open this PR? We're going to keep #3012 open.

Sounds good, I think this may have to change slightly to be compatible with a potential electrical redesign too.

**Pull request closed without merging**